draft/celery-kubernetes-operator.rst

```diff
@@ -0,0 +1,273 @@
+simply create an HPA object in creation/updation handlers and it'll take
+care of scaling relevant worker deployment automatically. HPA supports
+these two metrics out of the box. For custom metrics, we need to do
+additional work.
+
+Broker Queue Length(KEDA based autoscaling)
+-------------------------------------------
+
+Queue Length based scaling needs custom metric server for an HPA to
+work. `KEDA <https://keda.sh/docs/1.5/concepts/>`__ is a wonderful
+option because it is built for the same. It provides the
+`scalers <https://keda.sh/docs/1.5/scalers/>`__ for all the popular
+brokers(RabbitMQ, Redis, Amazon SQS) supported in Celery.
+
+KEDA provides multiple ways to be deployed on a Kubernetes cluster -
+Helm, Operator Hub and Yaml. Celery Operator can package KEDA along with
+itself for distribution.
+
+
+
+Deployment Strategy
+===================
+
+Probably the best way would be distribute a Helm Chart which packages
+CRD, controller and KEDA together(More to be explored here). We'll also
+support YAML apply based deployments.
+
+Additionally, Helm approach is extensible in the sense that we can
+package additional components like preferred broker(Redis, RMQ, SQS) as
+well to start with Celery on Kubernetes out of the box without much
+efforts.
+
+
+Motivation & Rationale
+======================
+
+Celery is one of the most popular distributed task queue system written
+in Python. Kubernetes is the de-facto standard for
+container-orchestration. We plan to write this operator to help manage
+celery applications gracefully and with ease on a Kubernetes cluster.
+
+Moreover, it'd be great to build this operator with Python. Kubernetes is
+written in golang. There is a good learning curve to understand
+internals and write(also maintain) an operator with Go.
+
+With the help of framework like KOPF, it'll be good to have Celery spearhead the Python
+ecosystem for developing production ready Kubernetes extensions. It'll
+motivate community to overcome the learning barrier and create useful
+libraries, tools and other operators while staying in Python ecosystem.
+
+
+Reference Implementation
+========================
+
+Github Repo - https://github.com/brainbreaker/Celery-Kubernetes-Operator
+Steps to try out are available in the Readme.
+
+Also a talk in EuroPython 2020 describing this POC implementation and demo - `Youtube <https://www.youtube.com/watch?v=MoVHxRZ1688&feature=youtu.be&t=9882>`__
+
+
+Want to Help?
+=============
+
+If you're running celery on a Kubernetes cluster, your inputs to how you
+manage applications will be valuable. You could contribute to the
+discussion `here <https://github.com/brainbreaker/Celery-Kubernetes-Operator/issues/12>`__.
+
+
+TODOs for Exploration
+=====================
+
+-  Helm chart to install the operator along with a broker of choice
+-  Add role based access control section for the operator
+-  Ingress Resource
+-  KEDA Autoscaling Implementation
```

Comment:
I'm happy to help with this one. Currently, we implemented `APIMetricScaler` in KEDA which allows users to use any arbitrary API for scaling (for example Flower API): https://keda.sh/docs/2.0/scalers/metrics-api/

I personally had a good experience with using KEDA to scale Celery.